mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp

```diff
@@ -1289,28 +1351,108 @@ getUntiledConsumerFromSlice(tensor::ParallelInsertSliceOp candidateSliceOp) {
   return getConsumerFromUses(resultingValue, containingOp->getBlock());
 }
 
-/// This utility currently checks whether the loop either :-
-/// 1. Yields exactly one result.
-/// 2. Has consumer op as its first user and other users to be in the same
-/// containing block as that of consumer op's. Currently we clone the loop op
-/// right before the consumer op in order to maintain a valid def-use chain.
-/// This utility thus helps ensuring that no invalid IR is formed due to the
-/// same.
+/// This utility currently checks whether the first userOp of loop is NOT before
+/// the last defineOp of consumer. Currently we clone the loop op right before
+/// a certain op in order to maintain a valid def-use chain. This utility thus
+/// helps ensuring that no invalid IR is formed due to the same. E.g.
+///
+/// ```
+/// %0 = scf.for() {
+///
+/// }
+/// ...
+/// %1 = firstUserOfLoop(%0)
+/// ...
+/// %2 = lastDefOfConsumer
+/// ...
+/// %3 = consumerOp(%2)
+/// ```
+///
+/// If the `firstUserOfLoop`is before `lastDefOfConsumer`, then it would be
+/// invalid to clone the loop op right before the `firstUserOfLoop`:
+///
+/// ```
+/// %0:2 = scf.for() {
+///    %3 = tiledConsumerOp(%2)
+/// }
+/// %1 = firstUserOfLoop(%0)
+/// ...
+/// %2 = lastDefOfConsumer
+/// ```
+///
+/// To address this issue, this utility would double-check there is no user of
+/// `firstUserOfLoop` before `lastDefOfConsumer`. If so, moving
+/// `firstUserOfLoop` after `lastDefOfConsumer`. Then, it turns out valid as
+/// follow:
+///
+/// ```
+/// %2 = lastDefOfConsumer
+/// %0:2 = scf.for() {
+///    %3 = tiledConsumerOp(%2)
+/// }
+/// %1 = firstUserOfLoop(%0)
+/// ```
+///
+/// @param loopOp: loop operation
```

Comment:
This seems like a general utility (I wrote somethign similar in a downstream project). Can we make this a general utility function that does some code motion to allow for better handling of use-def change issues. 

This is what I wrote https://github.com/iree-org/iree/pull/17880/files#diff-5fd6606dd19a6f5af5348cfc711ae8582543452897390bb3320f8ead1281aeb0R324 . Maybe that can be made a general utility (and avoid cloning) to make this file less complex.

Reply:
> For later, but I would just use DominanceInfo here.

> Can we make this a general utility function that does some code motion to allow for better handling of use-def change issues.

> Maybe that can be made a general utility (and avoid cloning) to make this file less complex.

Sure, I agree. BTW, could you help to point out the concrete related code pieces in you PR? That would be useful to quickly grab what we can reuse.

Reply:
https://github.com/iree-org/iree/blob/16f7ebf3883fd319152518c067121fdabad00b03/compiler/src/iree/compiler/Dialect/Flow/Transforms/FuseHorizontalContractions.cpp#L324

Reply:
Thanks for navigation. After reading that piece of code, I am afraid that we can not directly reuse the logic at the moment for two major reasons:

1. You are moving up def operands while this patch aims to move use result.  
2. There exists assumption for your utility. E.g. in your use-case, `insertPoint` op is the first contraction op of horizontal group. You can get the backward slice that not dominates `insertPoint` op and then move them up. However, how do we ensure this moving is safe enough in more general scenarios except for horizontal cases? I.e. What if some of backward slice depends on the result of `insertPoint` op. Actually that is what this patch need to check.

IMO, I still agree that we can unify these two logic but perhaps not in this patch.